### ISSUE TITLE
Auth user can't be found by Admin SDK in emulated function

### ISSUE THREAD

**Issue by chrisjacobs91** [firebase/firebase-tools#2878]
### [REQUIRED] Environment info

**firebase-tools:** 8.16.2

**Platform:** macOS

### [REQUIRED] Test case

- Create an auth user in the emulator UI
- Create a firestore document with an ID that matches the newly created auth user's UID
- onCreate firestore triggered function has code:
```
const uid = context.params.id _(which checking in console, is correctly getting the document ID = auth user's UID)_
const userAuth = await admin.auth().getUser(uid);
```
- Function logs show following error:
```
FirebaseAuthError: There is no user record corresponding to the provided identifier.
    at FirebaseAuthError.FirebaseError [as constructor] (/Users/cj/Documents/sailties/Firebase/functions/node_modules/firebase-admin/lib/utils/error.js:42:28)
    at FirebaseAuthError.PrefixedFirebaseError [as constructor] (/Users/cj/Documents/sailties/Firebase/functions/node_modules/firebase-admin/lib/utils/error.js:88:28)
    at new FirebaseAuthError (/Users/cj/Documents/sailties/Firebase/functions/node_modules/firebase-admin/lib/utils/error.js:147:16)
    at /Users/cj/Documents/sailties/Firebase/functions/node_modules/firebase-admin/lib/auth/auth-api-request.js:518:15
    at /Users/cj/Documents/sailties/Firebase/functions/node_modules/firebase-admin/lib/auth/auth-api-request.js:1476:13
    at processTicksAndRejections (node:internal/process/task_queues:93:5)
    at async /Users/cj/Documents/sailties/Firebase/functions/lib/main.js:734:26 {
  errorInfo: {
    code: 'auth/user-not-found',
    message: 'There is no user record corresponding to the provided identifier.'
  },
  codePrefix: 'auth'
}
```

### [REQUIRED] Steps to reproduce

See above

### [REQUIRED] Expected behavior

The auth user object should have been returned

### [REQUIRED] Actual behavior

Error, auth user doesn't exist.

**Comment by samtstern** [comment]
@chrisjacobs91 are you using `firebase-admin` version `9.4.0` or later?  Also can you show the full code of your firestore trigger including how you initialize `firebase-admin`?

**Comment by chrisjacobs91** [comment]
Hi @samtstern - how do I check the firebase-admin version that I'm using in cloud functions?

This is my full function:

```
export const profileCreated = functions.firestore.document(`profileInfo/{uid}`).onCreate(async (snap, context) => {
  try {
    const uid = context.params.uid;
    console.log(`Profile created for user: ${uid}`);

    const data = snap.data();

    if(!data) {
      console.log(`Error: No document data`);
      return null;
    };

    // Send welcome email
    const userAuth = await admin.auth().getUser(uid); ---> Erroring at this point
    const userAuthdata = userAuth.toJSON();
    console.log(userAuthdata);
    
    const mailchimpSubscribingUser = {
      email: "<TBC once auth is working>",
      fullName: data.fullName
    };
    
    if(mailchimpSubscribingUser.email && mailchimpSubscribingUser.fullName) {
      await addUserToMailchimp("<SENSITIVE DATA>", mailchimpSubscribingUser);
      
      async function addUserToMailchimp(listId:string, user:{email:string, fullName:string}) {

        const mailchimp = require('@mailchimp/mailchimp_marketing');
        mailchimp.setConfig({
          apiKey: "<SENSITIVE DATA>",
          server: "<SENSITIVE DATA",
        });

        const response = await mailchimp.lists.addListMember(listId, {
          email_address: user.email,
          status: "subscribed",
          merge_fields: {
            FULLNAME: user.fullName,
          }
        });
  
        console.log(`Successfully added contact as an audience member. The contact's id is ${response.id}.`);
  
        return
      };
    };

    return null;

  } catch(error) {
    console.log(error)
    return null;
  };

});
```

**Comment by samtstern** [comment]
@chrisjacobs91 it's the version listed in your `package.json`

**Comment by chrisjacobs91** [comment]
ah, of course, it's 9.4.1.

**Comment by samtstern** [comment]
@chrisjacobs91 I can't reproduce this.  I did a brand new `firebase init` and wrote the following simple functions:

```js
const functions = require("firebase-functions");
const admin = require("firebase-admin");
admin.initializeApp();

exports.triggerFlow = functions.https.onRequest(async (request, response) => {
  // Create a user
  const uid = `user${new Date().getTime()}`;

  console.log("Creating user", uid);
  await admin.auth().createUser({
    uid
  });

  // Create a matching Firestore document
  await admin.firestore().doc(`profileInfo/${uid}`).set({
    uid
  });

  response.json({
    uid
  });
});

exports.profileCreated = functions.firestore
  .document(`profileInfo/{uid}`)
  .onCreate(async (snap, context) => {
    const uid = context.params.uid;
    console.log(`Profile created for user: ${uid}`);

    const data = snap.data();

    if (!data) {
      console.log(`Error: No document data`);
      return null;
    }

    // Retrieve the user
    const userAuth = await admin.auth().getUser(uid);
    const userAuthdata = userAuth.toJSON();
    console.log(userAuthdata);
  });
```

Then I trigger the HTTP function to kick off the flow:
```
$ http http://localhost:5001/fir-dumpster/us-central1/triggerFlow
HTTP/1.1 200 OK
connection: keep-alive
content-length: 27
content-type: application/json; charset=utf-8
date: Tue, 01 Dec 2020 21:46:26 GMT
etag: W/"1b-nkbjqAwdvIO0zVxNwiiKxCBGG5I"
x-powered-by: Express

{
    "uid": "user1606859186120"
}
```

And I see the right logs in the terminal:

```
i  functions: Beginning execution of "triggerFlow"
>  Creating user user1606859186120
i  functions: Finished "triggerFlow" in ~1s
i  functions: Beginning execution of "profileCreated"
>  Profile created for user: user1606859186120
>  { uid: 'user1606859186120',
>    email: undefined,
>    emailVerified: false,
>    displayName: undefined,
>    photoURL: undefined,
>    phoneNumber: undefined,
>    disabled: false,
>    metadata:
>     { lastSignInTime: 'Tue, 01 Dec 2020 21:46:26 GMT',
>       creationTime: 'Tue, 01 Dec 2020 21:46:26 GMT' },
>    passwordHash: undefined,
>    passwordSalt: undefined,
>    customClaims: undefined,
>    tokensValidAfterTime: undefined,
>    tenantId: undefined,
>    providerData: [] }
i  functions: Finished "profileCreated" in ~1s
```

For completeness, here's my `firebase.json`:
```json
{
  "firestore": {
    "rules": "firestore.rules",
    "indexes": "firestore.indexes.json"
  },
  "emulators": {
    "auth": {
      "port": 9099
    },
    "functions": {
      "port": 5001
    },
    "firestore": {
      "port": 8080
    },
    "ui": {
      "enabled": true
    }
  }
}
```

So it has to be something in your local setup, the emulators seem to be working fine.

**Comment by chrisjacobs91** [comment]
@samtstern I went away but left the emulator running and came back about 3 hours later and it was working fine.

This morning I've restarted the emulator and I'm getting the original issue again. So it seems like there is some sort of delay between me creating an auth user and the Admin SDK being able to detect it.

I presume this didn't happen to you?

**Comment by chrisjacobs91** [comment]
@samtstern I quit the emulator and restarted and this time everything is working straight away...I didn't change anything - but happy with the result...!

Will close this and hope for the best!

**Comment by samtstern** [comment]
Thanks for following up, let us know if this issue comes back.

**Comment by leyenda** [comment]
I was having this issue following the docs where it states that a projectId must be setted

`admin.initializeApp({ projectId: "your-project-id" });`

Using that the auth was unable to list the users.

If i use `admin.initializeApp();` the admin sdk is now able to list the users created in the auth emulator.

**Comment by kantasv** [comment]
Hi, I am still suffering from the same situation.

FYI, my environment:

```
CLI Version:   9.22.0
Platform:      darwin, macOS Montrey 12.0.1
Node Version:  v12.12.0
Env Overrides: FIRESTORE_EMULATOR_HOST
Initialization: admin.initializeApp()
```

Any other people who have the same issue?

**Comment by bkendall** [comment]
Since this is a year old, could you open a new issue? That'd help for our tracking. Thanks!

**Comment by stansotn** [comment]
Ran into this issue when testing functions with `firebase-functions-test` in the following environment.
```
firebase-admin: v10.0.0
firebase-functions-test: v0.3.3
Platform: darwin, macOS Montrey 12.0.1
FIREBASE_DATABASE_EMULATOR_HOST="localhost:9000"
```

My app was initialized as follows:
```
import admin from "firebase-admin";
import test from "firebase-functions-test";

admin.initializeApp({
    databaseURL: "http://localhost:9000/?ns=ascii-docs-default-rtdb",
    projectId: "project-id",
  });

const wrapper = test({
  databaseURL: "http://localhost:9000/?ns=ascii-docs-default-rtdb",
  projectId: "project-id",
});
```

TLDR: The following config works (removed `projectId` from both initialization calls):
```
import admin from "firebase-admin";
import test from "firebase-functions-test";

admin.initializeApp({
    databaseURL: "http://localhost:9000/?ns=ascii-docs-default-rtdb",
  });

const wrapper = test({
  databaseURL: "http://localhost:9000/?ns=ascii-docs-default-rtdb",
});
```
The strange observation about this issue is that admin sdk connects to the auth emulator just fine, but when I call `admin.auth().listUsers()` it returns an empty list, even though auth emulator contains valid user accounts when I look at the firebase emulator suite (the web interface). Also, auth emulator works fine when I use `getAuth()` from `firebase/auth` web api v9 in the same test script.

As @leyenda mentioned, I removed `projectId: "project-id"` from both initializeApp and functions test wrapper, then calls to `admin.auth()` work as expected. Removing `projectId` only from the `initializeApp` does not help if `projectId` is mentioned in the `firebase-functions-test` initialization.

Smells like a bug to me.
